### ISSUE TITLE
Loader boundaries

### ISSUE THREAD

**Issue by ricokahler** [JustSift/ReSift#16]
**TL;DR** this is a proposal enables us to write code like this:

```js
import React from 'react';
import thingFetch from './thingFetch';
import { useData, withLoaderBoundary } from 'resift';

function Thing() {
  // thing will always be fetched
  const [thing] = useData(thingFetch, dispatch => {
    dispatch(thingFetch());
  });

  return <div>{thing.stuff}</div>
}

export default withLoaderBoundary(Thing);
```

ReSift will now handle dispatching fetches for you and (unlike `useFetch`) the data `useData` returns will always be defined and fetched like magic.

---

**Not TL;DR**

So I was reading [Dan Abramov's article on "algebraic effects"](https://overreacted.io/algebraic-effects-for-the-rest-of-us) and in that article he linked a [tweet/gist from Sebastian Markbåge](https://mobile.twitter.com/sebmarkbage/status/941214259505119232).

This gist implements a way to get values from promises without using `.then()` and without using `async`/`await` which is pretty fascinating/game changing.

Check it out:

usage:

```js

function getUserName(id) {
  var user = JSON.parse(fetchTextSync('/users/' + id));
  return user.name;
}

function getGreeting(name) {
  if (name === 'Seb') {
    return 'Hey';
  }
  return fetchTextSync('/greeting');
}

function getMessage() {
  let name = getUserName(123);
  return getGreeting(name) + ', ' + name + '!'; 
}

runPureTask(getMessage).then(message => console.log(message))
```

implementation:

```js
let cache = new Map();
let pending = new Map();

function fetchTextSync(url) {
  if (cache.has(url)) {
    return cache.get(url);
  }
  if (pending.has(url)) {
    throw pending.get(url);
  }
  let promise = fetch(url).then(
    response => response.text()
  ).then(
    text => {
      pending.delete(url);
      cache.set(url, text);
    }
  );
  pending.set(url, promise);
  throw promise;
}

async function runPureTask(task) {
  for (;;) {
    try {
      return task();
    } catch (x) {
      if (x instanceof Promise) {
        await x;
      } else {
        throw x;
      }
    }
  }
}
```

And then someone replied to this tweet implementing the same concept but using React with [error boundaries](https://reactjs.org/docs/error-boundaries.html):

![image](https://user-images.githubusercontent.com/10551026/64803353-94b25980-d55a-11e9-83f3-2e5e028713d7.png)

---

## What's going on here?

The code example above is getting values from promises "synchronously" via hacking the `throw` statement. When a component encounters these pseudo promises, it throws stopping the current execution. However, when the promise actually resolves, there's a loop that will re-try the execution. Since there is data, the flow will execute without throwing.

Since the React components are [pure/idempotent](https://overreacted.io/react-as-a-ui-runtime/#purity), we can have them partially re-execute without the worry of side-effects.

When you add React's [error boundaries](https://reactjs.org/docs/error-boundaries.html) with `componentDidCatch`, we can implement this idea of **"loader boundaries"**.

---

## Loader Boundaries: A game changing proposal

Loader boundaries are components that will prevent rendering of components with fetches until the values from those fetches become defined. What these boundaries will do is "catch" thrown fetches and prevent rendering until those fetches are finished.

This enables us to write hooks that will always assert that data is there and furthermore we can make these hooks calls dispatch fetches for data if there is none.

The goal of this proposal is to have this hook `useData` (tentative name) that will:

- dispatch the a request for the fetch if the data isn't there yet
- return the result of your fetch like it's already there and fallback to the closest loader boundary

```js
function Person({ personId }) {
  const personFetch = makePersonFetch(personId);

  // person will always be defined
  // this hook will also dispatch a request if there isn't an inflight one yet
  const [person] = useData(personFetch, dispatch => {
    dispatch(personFetch())
  });

  return <div>{person.name}</div>;
}

function App() {
  <LoaderBoundary fallback={<CircularProgress />}>
    <Person id="person123" />
  </LoaderBoundary>
}
```

No more `dispatch`ing, no more handling `null`. Just use the data.

## Concerns

- the implementation of the task runner above contains an infinite loop. this implementation explains the concept but should definitely not be the implementation we move forward with
- our current `Loader` (from skipper) allows us to wrap part of a component with a spinner but this pattern would force us to split up our components more because one component would stop rendering 
- These loader boundaries may unmount the component with `useData` calls if the fetch status becomes unknown suddenly. This may cause state to be reset too often.
- React will ship real suspense for data fetching where we don't need to hack the error boundaries. I think these proposed APIs will be compatible with suspense for data fetching but will probably require some refactors when we switch to that.

Let me know what you think!

**Comment by ricokahler** [comment]
So I implemented this and it yells at you for each object thrown in error boundaries so this is not an ideal solution at all lol

I'll close this until suspense for data fetches comes out haha
